File: app/XDM/XDM.Core/HttpServer/HttpServer.cs
using System;
using System.Net;
using System.Net.Sockets;
using System.Threading;
using TraceLog;

namespace XDM.Core.HttpServer
{
    public class NanoServer
    {
        private readonly TcpListener listener;
        public event EventHandler<RequestContextEventArgs>? RequestReceived;

        public NanoServer(int port) : this(IPAddress.Any, port) { }

        public NanoServer(IPAddress host, int port)
        {
            this.listener = new TcpListener(host, port);
        }

        public void Start()
        {
            listener.Start();
            while (true)
            {
                var tcp = listener.AcceptTcpClient();
                ProcessRequest(tcp);
            }
        }

        public void Stop()
        {
            try
            {
                this.listener.Stop();
            }
            catch { }
        }

        private void ProcessRequest(TcpClient tcp)
        {
            new Thread(() =>
            {
                try
                {
                    while (true)
                    {
                        var ctx = HttpParser.ParseContext(tcp);
                        this.RequestReceived?.Invoke(this, new RequestContextEventArgs(ctx));
                        if (!ctx.KeepAlive)
                        {
                            break;
                        }
                    }
                }
                catch (Exception ex)
                {
                    Log.Debug(ex, ex.Message);
                }
                finally
                {
                    try { tcp.Close(); } catch { }
                }
            }).Start();
        }
    }
}

File: xdm-app/src/main/java/xdm/integration/HttpServer.kt
package xdm.integration


import xdm.core.util.Logger
import java.io.IOException
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.net.Socket

class HttpServer(
    private val host: String,
    private val port: Int,
    private val requestListener: (ctx: RequestContext) -> Unit,
    private val onSuccess: () -> Unit, private val onFailure: () -> Unit,
) {
    private val serverSocket = ServerSocket()

    fun start() {
        Thread {
            try {
                serverSocket.bind(InetSocketAddress(host, port))
                onSuccess()
                while (!serverSocket.isClosed) {
                    process()
                }
            } catch (ex: Exception) {
                Logger.info(ex)
                onFailure()
            }
        }.start()
    }

    private fun process() {
        try {
            val socket = serverSocket.accept()
            if (!socket.inetAddress.isLoopbackAddress) {
                Logger.info("INTEGRATION", "Rejected non-loopback connection from ${socket.inetAddress.hostAddress}")
                socket.close()
                return
            }
            processRequest(socket)
        } catch (e: IOException) {
            if (!serverSocket.isClosed) Logger.info(e)
        }
    }

    fun stop() {
        try {
            serverSocket.close()
        } catch (e: IOException) {
            // ignore
        }
    }

    private fun processRequest(socket: Socket) {
        Thread {
            try {
                socket.use {
                    while (true) {
                        val ctx = HttpParser.parseContext(socket)
                        requestListener(ctx)
                        if (!ctx.keepAlive) {
                            break
                        }
                    }
                }
            } catch (_: HttpParser.ConnectionClosedException) {
                // Client closed an idle kept-alive connection.
            } catch (e: Exception) {
                Logger.info(e.message)
            }
        }.start()
    }
}
